src/dbnode/persist/fs/retriever.go

```diff
@@ -534,67 +567,62 @@ func (r *blockRetriever) Stream(
 	// the data. This means that even though we're returning nil for error
 	// here, the caller may still encounter an error when they attempt to
 	// read the data.
-	return req.toBlock(), nil
+	return nil
 }
 
-func (r *blockRetriever) StreamIndexChecksum(
+func (req *retrieveRequest) toBlock() xio.BlockReader {
+	return xio.BlockReader{
+		SegmentReader: req,
+		Start:         req.start,
+		BlockSize:     req.blockSize,
+	}
+}
+
+func (r *blockRetriever) Stream(
 	ctx context.Context,
 	shard uint32,
 	id ident.ID,
 	startTime time.Time,
+	onRetrieve block.OnRetrieveBlock,
 	nsCtx namespace.Context,
-) (block.StreamedChecksum, error) {
+) (xio.BlockReader, error) {
 	req := r.reqPool.Get()
-	req.shard = shard
-	// NB(r): Clone the ID as we're not positive it will stay valid throughout
-	// the lifecycle of the async request.
-	req.id = r.idPool.Clone(id)
-	req.start = startTime
-	req.blockSize = r.blockSize
-
-	req.streamReqType = streamIdxChecksumReq
-	req.resultWg.Add(1)
-
-	// Ensure to finalize at the end of request
-	ctx.RegisterFinalizer(req)
+	req.onRetrieve = onRetrieve
+	req.streamReqType = streamDataReq
 
-	// Capture variable and RLock() because this slice can be modified in the
-	// Open() method
-	r.RLock()
-	// This should never happen unless caller tries to use Stream() before Open()
-	if r.seekerMgr == nil {
-		r.RUnlock()
-		return block.EmptyStreamedChecksum, errNoSeekerMgr
+	onNoIDFn := func() {
+		// No need to call req.onRetrieve.OnRetrieveBlock if there is no data.
+		req.onRetrieved(ts.Segment{}, namespace.Context{})
 	}
-	r.RUnlock()
 
-	idExists, err := r.seekerMgr.Test(id, shard, startTime)
+	err := r.streamRequest(ctx, req, shard, id, startTime, onNoIDFn, nsCtx)
 	if err != nil {
-		return block.EmptyStreamedChecksum, err
+		return xio.EmptyBlockReader, err
 	}
 
-	// If the ID is not in the seeker's bloom filter, then it's definitely not on
-	// disk and we can return immediately.
-	if !idExists {
-		return block.EmptyStreamedChecksum, nil
-	}
+	// The request may not have completed yet, but it has an internal
+	// waitgroup which the caller will have to wait for before retrieving
+	// the data. This means that even though we're returning nil for error
+	// here, the caller may still encounter an error when they attempt to
+	// read the data.
+	return req.toBlock(), nil
+}
 
-	reqs, err := r.shardRequests(shard)
+func (r *blockRetriever) StreamIndexChecksum(
+	ctx context.Context,
+	shard uint32,
+	id ident.ID,
+	startTime time.Time,
+	nsCtx namespace.Context,
+) (block.StreamedChecksum, error) {
+	req := r.reqPool.Get()
+	req.streamReqType = streamIdxChecksumReq
+
+	err := r.streamRequest(ctx, req, shard, id, startTime, nil, nsCtx)
```

Comment:
Hm if this does nothing based on when `!idExists` the request will never be fulfilled no? Since it's not added to the list of outstanding requests to be serviced by the background goroutine.